### PR TITLE
Add: Distrubuted id allocator

### PR DESCRIPTION
**Pull request by ActivePeter**
The design issue: https://github.com/datenlord/datenlord/issues/379

### CODE REVIEW

**Comment by xiaguan** [comment]
MetaTxn() does not require a cancel() interface as it is designed for the read-modify-writeback model. If you receive an unwanted value during a get() operation, you can simply return an Error or use KVEngine::get().

**Comment by ActivePeter** [comment]
> `MetaTxn` does not require a `cancel()` interface as it is designed for the read-modify-writeback model. If you receive an unwanted value during a `get()` operation, you can simply return an Error or use `KVEngine::get()`.

The purpose here is to make retry_txn macro more convenient. In some cases we want to break the retry loop but we don't want to commit the txn, then we can call this function. Just check the code used this function.

**Comment by xiaguan** [comment]
ID allocation shouldn't require locks.Remove the content related to lock impl.

File: src/async_fuse/memfs/kv_engine.rs [hunk 1]
```diff
@@ -20,66 +23,127 @@ pub enum ValueType {
...
+    /// # Panics
     /// Panics if `ValueType` is not `ValueType::Node`.
     pub async fn into_s3_node<S: S3BackEnd + Send + Sync + 'static, K: KVEngine + 'static>(
         self,
         meta: &S3MetaData<S, K>,
     ) -> S3Node<S> {
         match self {
             ValueType::Node(node) => S3Node::from_serial_node(node, meta).await,
-            ValueType::DirEntry(_) | ValueType::INum(_) | ValueType::Attr(_) => {
+            _ => {
```
Comment (rogercloud) [inline src/async_fuse/memfs/kv_engine.rs:41]: We'd better list all the possible values. If we add new enum option later, we may handle it incorrectly as there's no error reporting for the wildcard_enum_match.

File: src/async_fuse/memfs/kv_engine.rs [hunk 2]
```diff
@@ -20,66 +23,127 @@ pub enum ValueType {
...
-    #[allow(dead_code)]
+    /// Turn the `ValueType` into `NextIdAllocateRangeBegin`.
+    /// # Panics
+    /// Panics if `ValueType` is not `ValueType::NextIdAllocateRangeBegin`.
+    #[allow(clippy::wildcard_enum_match_arm)]
     #[must_use]
+    pub fn into_next_id_allocate_range_begin(self) -> u64 {
+        match self {
+            ValueType::NextIdAllocateRangeBegin(begin) => begin,
+            _ => panic!("expect ValueType::NextIdAllocateRangeBegin but get {self:?}"),
```
Comment (rogercloud) [inline src/async_fuse/memfs/kv_engine.rs:55]: The same as above.

File: src/async_fuse/memfs/kv_engine.rs [hunk 3]
```diff
@@ -20,66 +23,127 @@ pub enum ValueType {
...
-#[derive(Debug)]
 /// The `KeyType` is used to locate the value in the distributed K/V storage.
 /// Every key is prefixed with a string to indicate the type of the value.
 /// If you want to add a new type of value, you need to add a new variant to the enum.
 /// And you need to add a new match arm to the `get_key` function , make sure the key is unique.
 #[allow(dead_code)]
-pub enum KeyType {
-    /// INum -> SerialNode
+#[derive(Debug, Eq, PartialEq)]
+pub enum KeyType<'a> {
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:78]: It's for str ref.

File: src/async_fuse/memfs/kv_engine.rs [hunk 4]
```diff
@@ -20,66 +23,127 @@ pub enum ValueType {
...
+impl KeyType<'_> {
     /// Get the key in bytes.
+    #[must_use]
     pub fn get_key(&self) -> Vec<u8> {
         match *self {
-            KeyType::INum2Node(ref i) => format!("I{i}").into_bytes(),
-            KeyType::INum2DirEntry(ref i) => format!("D{i}").into_bytes(),
-            KeyType::Path2INum(ref p) => format!("P{p}").into_bytes(),
-            KeyType::INum2Attr(ref i) => format!("A{i}").into_bytes(),
+            KeyType::INum2Node(ref i) => format!("NodeI{i}").into_bytes(),
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:133]: This is just an encoding problem, if we use one litter prefix, there‘s only 26*2 key prefixes can be used

File: src/async_fuse/memfs/kv_engine.rs [hunk 5]
```diff
@@ -193,6 +258,9 @@ impl MetaTxn for EtcdTxn {
     }
 
     async fn commit(&mut self) -> DatenLordResult<bool> {
+        if self.version_map.is_empty() && self.buffer.is_empty() {
```
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:261]: Just for improving code completeness.

**Comment by ActivePeter** [comment]
> 

Why?

**Comment by ActivePeter** [comment]
> ID allocation shouldn't require locks.Remove the content related to lock impl.

Why?

File: src/async_fuse/memfs/kv_engine.rs (hunk 1 above)
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:41]: These functions are for turning a common type (ValueType) into one certained type (S3_Node), if later we implement more types, would that be too long or weird? And actually we always are certain a key type is binded with a value type.

File: src/async_fuse/memfs/kv_engine.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:261]: We don't need to commit a transaction if there's no operation.

File: src/async_fuse/memfs/kv_engine.rs (hunk 1 above)
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:41]: Added comment for this.

File: src/async_fuse/memfs/kv_engine.rs (hunk 2 above)
Comment (ActivePeter) [reply src/async_fuse/memfs/kv_engine.rs:55]: Added comment for this in code.

**Review by xiaguan — APPROVED** [review]

**Pull request merged**